rust: Migrate zbus v1 to V5

cathay4t opened the pull request:
Fedora 42 is removing support of zbus V1 and we are suggested to use V4
or V5.

The zbus V4 does not have Clone trait for OwnedValue which require
massive changes to our code base. Hence using zbus V5.

The zbus V5 need at least 1.77 to compile as it use a nightly feature of
rust https://github.com/rust-lang/rust/issues/91345 which only become
stable in rust 1.77.

It is impossible to compile this project in Rust 1.66 now, removing
CI test for compiling on Rust 1.66. Downstream release should
considering using Rust 1.80+ to compile this project.

ihuguet commented on rust/src/lib/nm/nm_dbus/dbus.rs:
```diff
@@ -4,12 +4,10 @@ use std::collections::HashMap;
 
 use super::{
     connection::{NmConnection, NmConnectionDbusValue},
-    dbus_proxy::{
-        NetworkManagerDnsProxy, NetworkManagerProxy, NetworkManagerSettingProxy,
-    },
```
You have moved these proxies from dbus_proxy but didn't removed them from dbus_proxy.

cathay4t replied:
Thanks. I forgot the comment on this. This is required because zbus V5 is generating private `NetworkManagerDnsProxy` instead of `pub(crate)` now.

ihuguet replied:
There is no need to add the comment, then, this explanation is enough.

ihuguet commented on rust/src/lib/nm/nm_dbus/gen_conf/keyfile.rs:
```diff
@@ -60,7 +60,7 @@ fn zvariant_value_to_string(
         zvariant::Value::Array(a) => {
             let mut ret = String::new();
-            for item in a.get() {
+            for item in a.iter() {
```
Is this change right? Is it doing the same thing? I assume that zbus-v1 chose a weird name for returning the iterator, but just to be sure.

cathay4t replied:
It is correct. The old zvariant 2.10 has `pub fn get(&self) -> &[Value<'a>]`  but changed to `pub fn get<V>(&'a self, idx: usize) -> Result<Option<V>>` in 5.1.

Our original use is expecting a iterator auto-generated from a slice. So changing it to iter() is correct.

ihuguet commented on rust/src/lib/nm/nm_dbus/mod.rs:
```diff
@@ -5,8 +5,6 @@ mod connection;
 mod convert;
 #[cfg(feature = "query_apply")]
 mod dbus;
-#[cfg(feature = "query_apply")]
-mod dbus_proxy;
```
Should dbus_proxy file be deleted?

cathay4t replied:
Yes. Fixed.

ihuguet reviewed (changes requested):
Apart from the comments, LGTM. However, I think that Fedora is not doing the right thing. Zbus-v1 is only 4 years old and the update is not trivial.

cathay4t commented:
New patch uploaded with `dbus_proxy.rs` removed and added comment explaining why `zvariant::proxy()` should be sit with its consumer.

cathay4t committed:
rust: Migrate zbus v1 to V5

Fedora 42 is removing support of zbus V1 and we are suggested to use V4
or V5.

The zbus V4 does not have Clone trait for OwnedValue which require
massive changes to our code base. Hence using zbus V5.

The zbus V5 need at least 1.77 to compile as it use a nightly feature of
rust https://github.com/rust-lang/rust/issues/91345 which only become
stable in rust 1.77.

It is impossible to compile this project in Rust 1.66 now, removing
CI test for compiling on Rust 1.66. Downstream release should
considering using Rust 1.77+ to compile this project.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
Fedora has released rust-zbus V5. Please review the PR.

dustymabe commented:
> Fedora has released rust-zbus V5. Please review the PR.

Yes. Please do

liangwen12year commented:
The checkpoint management code (`nm_checkpoint_create`, `nm_checkpoint_rollback`, etc.) is now asynchronous, but the retry logic might not work as expected in an async context. 

For example:

```
if let Err(e) = nm_api.device_delete(&nm_dev.obj_path).await {
    log::debug!("Failed to delete interface {:?}", e);
}
```

If the deletion fails, the error is logged, but the function continues without retrying or propagating the error.

Can we add proper retry logic for critical operations like checkpoint creation and deletion. For example:

```
with_retry(RETRY_COUNT, RETRY_INTERVAL, || async {
    nm_api.device_delete(&nm_dev.obj_path).await
}).await?;
```

cathay4t commented:
> The checkpoint management code (`nm_checkpoint_create`, `nm_checkpoint_rollback`, etc.) is now asynchronous, but the retry logic might not work as expected in an async context.
> 
> For example:
> 
> ```
> if let Err(e) = nm_api.device_delete(&nm_dev.obj_path).await {
>     log::debug!("Failed to delete interface {:?}", e);
> }
> ```
> 
> If the deletion fails, the error is logged, but the function continues without retrying or propagating the error.
> 
> Can we add proper retry logic for critical operations like checkpoint creation and deletion. For example:
> 
> ```
> with_retry(RETRY_COUNT, RETRY_INTERVAL, || async {
>     nm_api.device_delete(&nm_dev.obj_path).await
> }).await?;
> ```

Even before this patch, we do not retry on device_delete but use verification retry. Please check again.

dustymabe commented:
Thanks @cathay4t - can we get this backported to the Fedora 43 and 42 RPMs ?

cathay4t commented:
> Thanks @cathay4t - can we get this backported to the Fedora 43 and 42 RPMs ?

Working on it.

cathay4t commented:
@dustymabe https://bodhi.fedoraproject.org/updates/FEDORA-2025-1901d00ff4 https://bodhi.fedoraproject.org/updates/FEDORA-2025-4f5d0e6cb5 created

dustymabe commented:
Thank You!
